flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sorted/state/SingleKeyKeyedStateBackend.java

```diff
@@ -0,0 +1,260 @@
+import org.apache.flink.runtime.state.internal.InternalKvState;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.RunnableFuture;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A simple {@link CheckpointableKeyedStateBackend} which keeps values for a single key at a time.
+ *
+ * <p><b>IMPORTANT:</b> Requires the incoming records to be sorted/grouped by the key. Used in a BATCH style execution.
+ */
+class SingleKeyKeyedStateBackend<K> implements CheckpointableKeyedStateBackend<K> {
+	@SuppressWarnings("rawtypes")
+	private static final Map<Class<? extends StateDescriptor>, StateFactory> STATE_FACTORIES =
+		Stream.of(
+			Tuple2.of(ValueStateDescriptor.class, (StateFactory) SingleKeyValueState::create),
+			Tuple2.of(ListStateDescriptor.class, (StateFactory) SingleKeyListState::create),
+			Tuple2.of(MapStateDescriptor.class, (StateFactory) SingleKeyMapState::create),
+			Tuple2.of(AggregatingStateDescriptor.class, (StateFactory) SingleKeyAggregatingState::create),
+			Tuple2.of(ReducingStateDescriptor.class, (StateFactory) SingleKeyReducingState::create)
+		).collect(Collectors.toMap(t -> t.f0, t -> t.f1));
+
+	private K currentKey = null;
+	private final TypeSerializer<K> keySerializer;
+	private final List<KeySelectionListener<K>> keySelectionListeners = new ArrayList<>();
+	private final Map<String, State> states = new HashMap<>();
+	private final Map<String, KeyGroupedInternalPriorityQueue<?>> priorityQueues = new HashMap<>();
+	private final KeyGroupRange keyGroupRange;
+
+	public SingleKeyKeyedStateBackend(
+			TypeSerializer<K> keySerializer,
+			KeyGroupRange keyGroupRange) {
+		this.keySerializer = keySerializer;
+		this.keyGroupRange = keyGroupRange;
+	}
+
+	@Override
+	public void setCurrentKey(K newKey) {
+		if (!Objects.equals(newKey, currentKey)) {
+			notifyKeySelected(newKey);
+			for (State value : states.values()) {
+				((AbstractSingleKeyState<?, ?, ?>) value).clearAllNamespaces();
+			}
+			for (KeyGroupedInternalPriorityQueue<?> value : priorityQueues.values()) {
+				while (value.poll() != null) {
+					// remove everything for the key
+				}
+			}
+			this.currentKey = newKey;
+		}
+	}
+
+	@Override
+	public K getCurrentKey() {
+		return currentKey;
+	}
+
+	@Override
+	public TypeSerializer<K> getKeySerializer() {
+		return keySerializer;
+	}
+
+	@Override
+	public <N, S extends State, T> void applyToAllKeys(
+			N namespace,
+			TypeSerializer<N> namespaceSerializer,
+			StateDescriptor<S, T> stateDescriptor,
+			KeyedStateFunction<K, S> function) {
+		throw new UnsupportedOperationException("applyToAllKeys() is not supported in BATCH execution mode.");
```

Comment:
I'm wondering about the name here. If we always refer to `BATCH execution mode` in the exceptions, maybe the state backend should be called `BatchExecutionStateBackend`. Or do we think it could be used for any other purpose.

Reply:
Personally I can't think of other scenarios...

Reply:
Will rename the classes.